### PR TITLE
Enable Huggingface to Maxtext checkpoint conversion 

### PR DESCRIPTION
**Pull request by YixuanWang-99**
# Description

To enable the checkpoint conversion from Huggingface to maxtext. 

- Added to_maxtext.py to perform the conversion from hugging face to maxtext
- Added gemma2_to_mt.sh as an example to utimately run to_maxtext.py and it's logits check test

# Tests

We can test the converted checkpoint with the original HF checkpoint by using forward_pass_logit_checker.py. 
```
python3 -m "MaxText.tests.forward_pass_logit_checker" \
    "MaxText/configs/base.yml" \
    tokenizer_path=${TOKENIZER_PATH} \
    load_parameters_path="${OUTPUT_BASE_DIR}/0/items"\
    run_name=forward_pass_test_${MODEL_NAME}\
    per_device_batch_size=${PER_DEVICE_BATCH_SIZE} \
    model_name=${MODEL_NAME} \
    max_prefill_predict_length=4 \
    max_target_length=4 \
    dataset_type=synthetic \
    scan_layers=${scan_layers} \
    attention=dot_product \
    --max_kl_div=0.015 \
    --run_hf_model=True \
    --hf_model_path=${HF_MODEL_ID} \
```

The to_maxtext.py is tested with gemma2-2b, qwen3-06b, qwen3-4b, and qwen3-8b. 
Conversion of gemma-2-2b [check](https://screenshot.googleplex.com/5L4cVWuEqah3sCk).
Conversion of qwen3-0.6b [check](https://screenshot.googleplex.com/9jDDSD5ySoQrsSA)
Conversion of qwen3-4b [check](https://screenshot.googleplex.com/3SvSgnTsDNx58Zk)
Conversion of qwen3-8b [check](https://screenshot.googleplex.com/Ac6rfi8LEHnPFRK)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  hf_token = config.hf_access_token
+  # Load Hugging Face model, config, and state_dict
+  max_logging.log(f"Loading Hugging Face model: {model_id}...")
+  hf_config_obj = AutoConfig.from_pretrained(model_id, token=hf_token)
+  hf_model = AutoModelForCausalLM.from_pretrained(model_id, token=hf_token)
+  hf_state_dict_numpy = hf_model.state_dict()
+  for k, v in hf_state_dict_numpy.items():
+    hf_state_dict_numpy[k] = v.numpy()
+  del hf_model
+  max_logging.log("Hugging Face model loaded and converted to NumPy.")
```
Comment (shralex) [inline MaxText/utils/ckpt_conversion/to_maxtext.py]: HuggingFace one word

File: MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+  config = pyconfig.initialize(argv)
+  # check the supported model ids
+  if config.model_name not in HF_IDS:
+    raise ValueError(f"Unsupported model name: {config.model_name}. Supported models are: {list(HF_IDS.keys())}")
+
+  model_id = HF_IDS[config.model_name]
+  max_utils.print_system_information()
+  if not config.base_output_directory:
+    output_directory = f"{os.getcwd()}/mt_output"
```
Comment (shralex) [inline MaxText/utils/ckpt_conversion/to_maxtext.py]: how about this instead: os.path.join("tmp", config.run_name, "")
(if you like it please update other occurances of mt_output)

File: MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_mt.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+python3 -m "MaxText.utils.ckpt_conversion.to_maxtext" \
+  "MaxText/configs/base.yml" \
+  model_name="${MODEL_NAME}" \
+  base_output_directory="${OUTPUT_BASE_DIR}" \
+  per_device_batch_size="${PER_DEVICE_BATCH_SIZE}" \
+  run_name="run_to_mt" \
+  async_checkpointing="${ASYNC_CHECKPOINTING}" \
+  scan_layers="${SCAN_LAYERS}" 
+
+sleep 15
```
Comment (shralex) [inline MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_mt.sh]: lets remove this ?

File: MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 1 above)
Comment (YixuanWang-99) [reply MaxText/utils/ckpt_conversion/to_maxtext.py]: Fixed

File: MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 2 above)
Comment (YixuanWang-99) [reply MaxText/utils/ckpt_conversion/to_maxtext.py]: Updated. to_huggingface.py is updated as well

File: MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_mt.sh (hunk 3 above)
Comment (YixuanWang-99) [reply MaxText/utils/ckpt_conversion/examples/convert_gemma2_to_mt.sh]: Done

File: MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+  if checkpoint_manager is not None:
+    if save_checkpoint(checkpoint_manager, 0, final_save_state):
+      max_logging.log("saved a checkpoint at step 0")
+    # Upon preemption, exit when and only when all ongoing saves are complete.
+    if checkpoint_manager.reached_preemption(0):
+      checkpoint_manager.wait_until_finished()
+      sys.exit()
+
+  max_logging.log("Conversion complete.")
```
Comment (hengtaoguo) [inline MaxText/utils/ckpt_conversion/to_maxtext.py]: Conversion complete, checkpoint saved to {config.base_output_directory}?

**Review by hengtaoguo — APPROVED** [review]

File: MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 4 above)
Comment (YixuanWang-99) [reply MaxText/utils/ckpt_conversion/to_maxtext.py]: Done
